### DISCUSSION TITLE
Any Image File Format to PNG Image Processing and/or PNG Color Type Change

### DISCUSSION THREAD

**Discussion by BryansBusiness** [ImageMagick/ImageMagick#6687] [Digital Image Processing]
Hello ladies and gentlemen,

My name is Bryan Rodriguez and I have a very important question for you all. Currently I am working on a project that it's main goal is to process images and spew them out as a PNG in indexed color type (using 1 bit channel *black). Originally the project started off on a IOS SDK and I was trying to use Apples Core Foundation library for the image processing, I was able to provide any file format to their routines and it would give me a PNG formatted image but I was unable to configure the color type. Now this is basically the most important part as my applications send these images to a thermal printer so it really only can handle/deal with black. I am on the fence right now as I could either write the code for my IOS SDK or straight into the printers firmware, but I have really no idea where I would begin; I have read some of ImageMagick's docs and it seems like SetImageColor might be the routine that would be useful to me. I couldn't help but reach out for additional information since I am not too familiar with the whole repo, I would appreciate some insight and maybe pointers on things I may be overlooking. Any help is appreciated. I will provide below 2 images, both in PNG but different color modes so you may see what I am referring to. The one named LemurIndexed is the working format and the LemurRGB is what most PNG's color type get defaulted to. Thank you everyone and I hope to hear from you all soon!
![LemurIndexed](https://github.com/ImageMagick/ImageMagick/assets/92451674/ba52294b-f874-4605-a11a-bedd6e1cd64c)
![LemurRGB](https://github.com/ImageMagick/ImageMagick/assets/92451674/55167c0c-e3fe-4332-b416-309d1eb45175)

**Comment by snibgo** [comment]
The forum software renames images. Which image is which?

I'm not sure what you are asking. If you want to read the second image (which I have named `lemur2.png`) and write it as a palette PNG file with black and white only:
```
magick lemur2.png png8:l1.png
```
Note the `png8:` prefix to the output filename.

**Reply by BryansBusiness** [reply]
Pardon for the confusion. The first picture is the appropriate color mode for my printer which is indexed using black and white only. The second one is in RGB color mode. My question is a little more technical than just using the magick commands. I am in search of the actual function so I may see how the code would look like for such a task. So imagine what my scenario would look like is I have a jpg and I need to convert it into a png with indexed color type using only black and white. So I would need routines to pass that jpg formatted data convert it into png formatted data and make sure it is in indexed color type using black and white only. Please feel free to ask me additional questions and thank you for you quick response.

**Reply by fmw42** [reply]
You can convert from any color image to b/w PNG indexed color (PNG8:) as follows as one way in Imagemagick 8

`magick image.jpg -monochrome PNG8:result.png`



Example:

![logo](https://github.com/ImageMagick/ImageMagick/assets/13458218/1371a099-1c08-4dff-8259-c0253f8c68c0)

`magick logo.jpg -monochrome PNG8:logo_result.png`

![logo_result](https://github.com/ImageMagick/ImageMagick/assets/13458218/f485afb4-dc72-4eac-9376-6194d296915b)


Is this what you want?  Or do you need an image for each of the up to 256 values in the 8-bit color map?

**Reply by BryansBusiness** [reply]
Hello everyone. Hopefully with this comment you guys can understand a little more what I am trying to achieve. So below I will be sending a lot of images and this time I will make sure to comment them appropriately. Thank you @fmw42 and @snibgo for your comments so far. @fmw42 the image you converted to indexed above is technically in the right color mode but not with the correct bit color map/ bit depth. 

This is what a PNG file is supposed to look like and I have used boxes with red outlines to highlight what is most important to us. The 08 here represents bits per pixel per channel (AKA bit depth) and the 02 represents color type equal to RGB in this example. 
![PNGHeader](https://github.com/ImageMagick/ImageMagick/assets/92451674/8e6257a8-ca45-44ff-85eb-7e1870c46db1)

This is the binary of the first picture I sent a while ago (the only working format for my printer). As you can see 01 represents it only has 1 bit per pixel per channel(AKA 1 bit depth) which is only black and the 03 means it's in indexed color mode. 
![PNGBinary](https://github.com/ImageMagick/ImageMagick/assets/92451674/86aafd96-1e08-424c-b6a7-df0d85a3f60c)

Now this is the binary of the second picture I sent last time. As you can see it has a 08 meaning it has a bit depth of 8 and a 02 meaning it is in RGB color mode. 
![PNGRGBBinary](https://github.com/ImageMagick/ImageMagick/assets/92451674/b0b20acd-d231-4fe6-b75c-401c76bc1adb)

Now this screen shot is the binary of the image @fmw42 sent in the last comment as you can see it is in indexed mode shown by the 03 but it still has a bit depth/bit color map of 8 (usually meaning different shades of gray) 
<img width="530" alt="ImageMagickBinary" src="https://github.com/ImageMagick/ImageMagick/assets/92451674/295f30b3-7998-4687-aa44-195b587b9dcc">

This image is a screenshot of GIMPS options when converting a PNG to indexed and as you can see for the image to be formatted in the way I need it, I have to check off the option of only using black and white (1 bit) palette. I am providing this image because currently it is the only way I have been able to process and convert images to the correct format to be used with our printers.
<img width="375" alt="GimpMenu" src="https://github.com/ImageMagick/ImageMagick/assets/92451674/56183a35-e2d3-43c3-9264-737c726a6288">

I hope this has been a useful explanation of my needs and would appreciate a lot further assistance. Hope to hear from you guys soon. 

Best Regards
Bryan Rodriguez

**Reply by fmw42** [reply]
One of the IM developers will need to comment if you think the color map is not correct.

**Reply by BryansBusiness** [reply]
@fmw42 Well I am not saying its incorrect just not what I am looking for. Do you know who I may tag specifically so they may see it?

**Reply by fmw42** [reply]
This is the way IM makes color maps. I do not think there is any way to change it, currently.  You can contact @dlemstra or @magick about it, but they always look at these questions. However, they are very busy and this may not be a high priority.  One of the other uses may know more about the code than I.

**Reply by BryansBusiness** [reply]
Thanks brother. I appreciate all your help. Hopefully they will have sometime to assist me soon.

**Comment by snibgo** [comment]
> Please feel free to ask me additional questions ...

What function do you use to write the file? Does it take the filename as an argument? If so, then prefix the filename with "PNG8:".

EDIT: For example:
```
        status = MagickWriteImage(magick_wand, "PNG8:newlemur.png");
```

**Reply by BryansBusiness** [reply]
Please refer to the comment above :)

**Comment by snibgo** [comment]
I don't think IM can make a "black and white (1 bit) palette".

**Comment by dlemstra** [comment]
Writing 1 bit is currently not support the minimum number of bits is 8. Support for this could be added in the future. You will probably get it earlier if someone would open a pull request to add support for this.
